### PR TITLE
New IEs and fields for GTPv2 message

### PR DESCRIPTION
**Pull request by UmakantKulkarni**
Added new IEs and fields for GTPv2 message as per 3GPP TS 29.274   V16.2.0 (2019-12) - http://www.3gpp.org/ftp//Specs/archive/29_series/29.274/29274-g20.zip

### CODE REVIEW

**Comment by UmakantKulkarni** [comment]
@guedou , Thanks for pointing them out.  I've fixed these errors and committed the changes.
Please approve the PR if you don't see any issues.!
```
$ sudo test/run_tests -P "load_contrib('gtp_v2')" -t test/contrib/gtp_v2.uts > /tmp/uk1
### Booting scapy...
### Starting tests...
### Loading: test/contrib/gtp_v2.uts
passed 5CCFC8E4 000.00s GTPHeader v2, basic instantiation
passed 65EBE70E 000.00s GTPV2EchoRequest, basic instantiation
passed AA51C76B 000.00s GTPV2CreateSessionRequest, basic instantiation
passed F9C6810E 000.00s GTPV2EchoRequest, dissection
passed 429B7CBE 000.00s GTPV2EchoResponse, dissection
passed 9CFECF19 000.00s GTPV2ModifyBearerRequest, dissection
passed 4A51BF5A 000.01s IE_IMSI, dissection
passed D4FEA35D 000.00s IE_IMSI, basic instantiation
passed E52DB4DC 000.00s IE_Cause, dissection
passed 0F5CA9D1 000.00s IE_Cause, basic instantiation
passed EBD4DDA0 000.00s IE_Cause, basic instantiation 2
passed 50F8C5D1 000.00s IE_Cause, basic instantiation 3
passed 19654406 000.00s IE_RecoveryRestart, dissection
passed 7A57FDB7 000.00s IE_RecoveryRestart, basic instantiation
passed 94FE45CE 000.01s IE_APN, dissection
passed 1EAA5031 000.00s IE_APN, basic instantiation
passed DD2D2887 000.01s IE_AMBR, dissection
passed 45FA5ED0 000.00s IE_AMBR, basic instantiation
passed FCEB3271 000.00s IE_EPSBearerID, dissection
passed 53FF4499 000.00s IE_EPSBearerID, basic instantiation
passed 58247B94 000.00s IE_IPv4, dissection
passed 0317E22D 000.00s IE_IPv4, basic instantiation
passed 603DF9D7 000.00s IE_MEI, dissection
passed 50EE5061 000.00s IE_MEI, basic instantiation
passed EA549337 000.00s IE_MSISDN, dissection
passed 167FBE48 000.00s IE_MSISDN, basic instantiation
passed EF0027FF 000.00s IE_Indication, dissection
passed 752AF271 000.00s IE_Indication, basic instantiation
passed 5F4A2709 000.00s IE_Indication, basic instantiation 2
passed E5AD13E3 000.01s IE_PCO, dissection
passed 087BE85E 000.00s IE_PCO, basic instantiation
passed 60740CB5 000.00s IE_PAA, dissection
passed 90342883 000.00s IE_PAA, basic instantiation
passed CD8834C2 000.01s IE_Bearer_QoS, dissection
passed 2668574E 000.00s IE_Bearer_QoS, basic instantiation
passed 0551DF73 000.01s IE_RAT, dissection
passed 0B681092 000.00s IE_RAT, basic instantiation
passed 0AE4B86B 000.00s IE_ServingNetwork, dissection
passed 4DD7450E 000.00s IE_ServingNetwork, basic instantiation
passed DFB109BF 000.01s IE_ULI, dissection
passed A6A6BEEB 000.00s IE_ULI, basic instantiation
passed 57C935DB 000.00s IE_UCI, dissection
passed A53105F1 000.00s IE_UCI, basic instantiation
passed AB6DF1DB 000.01s IE_FTEID, dissection
passed 90D049F4 000.00s IE_FTEID, basic instantiation
passed 112E74BF 000.01s IE_BearerContext, dissection
passed CEF65F3B 000.00s IE_BearerContext, basic instantiation
passed 667A4D73 000.00s IE_ChargingID, dissection
passed A1535DB0 000.00s IE_ChargingID, basic instantiation
passed 07429366 000.01s IE_ChargingCharacteristics, dissection
passed D0A5FE91 000.00s IE_ChargingCharacteristics, basic instantiation
passed CF5A04A3 000.01s IE_PDN_type, dissection
passed F69821FB 000.00s IE_PDN_type, basic instantiation
passed A65C02FB 000.01s IE_UE_Timezone, dissection
passed 8CDCCEFC 000.00s IE_UE_Timezone, basic instantiation
passed E3712C0E 000.00s IE_UE_Timezone, basic instantiation
passed F03C04DA 000.00s IE_Port_Number, dissection
passed 2B7EC740 000.00s IE_Port_Number, basic instantiation
passed 0C09C023 000.00s IE_APN_Restriction, dissection
passed 46E822D5 000.00s IE_APN_Restriction, basic instantiation
passed 59382080 000.02s IE_SelectionMode, dissection
passed 56548512 000.00s IE_SelectionMode, basic instantiation
passed 896BAB27 000.01s IE_MMBR, dissection
passed AFF3EE0B 000.00s IE_MMBR, basic instantiation
passed 713402DB 000.00s GTPHeader answers to not GTPHeader instance
passed B553F724 000.00s GTPHeader post_build
passed 23030A61 000.00s GTPHeader hashret
passed 1A5D25DB 000.00s IE_NotImplementedTLV
passed 9A6A16FA 000.00s IE_PrivateExtension
Campaign CRC=2A1AE9ED in 000.17s SHA=F5278DCDF9FC2856947873103CFA87E94B33C54F
PASSED=69 FAILED=0
### Writing output...
```

File: scapy/contrib/gtp_v2.py [hunk 1]
```diff
@@ -456,6 +466,22 @@ class IE_ULI(gtp.IE_Base):
...
+                   BitField("instance", 0, 4),
+                   gtp.TBCDByteField("MCC", "", 2),
+                   gtp.TBCDByteField("MNC", "", 1),
+                   BitField("SPARE", 0, 5),
+                   BitField("CSG_ID", 0, 27),
+                   BitField("AccessMode", 0, 2),
+                   BitField("SPARE", 0, 4),
+                   BitField("LCSG", 0, 1),
+                   BitField("CMI", 0, 1)]
+
```
Comment (p-l-) [inline scapy/contrib/gtp_v2.py]: Please be careful: avoid trailing spaces. This causes the flake8 fail.

```suggestion

```

**Comment by UmakantKulkarni** [comment]
@p-l- Looks good now?

File: scapy/contrib/gtp_v2.py [hunk 2]
```diff
@@ -603,7 +645,7 @@ class IE_IMSI(gtp.IE_Base):
 class IE_Cause(gtp.IE_Base):
     name = "IE Cause"
     fields_desc = [ByteEnumField("ietype", 2, IEType),
-                   ShortField("length", 6),
+                   ShortField("length", None),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py]: Why are you changing this value?

File: scapy/contrib/gtp_v2.py [hunk 3]
```diff
@@ -616,7 +658,7 @@ class IE_Cause(gtp.IE_Base):
 class IE_RecoveryRestart(gtp.IE_Base):
     name = "IE Recovery Restart"
     fields_desc = [ByteEnumField("ietype", 3, IEType),
-                   ShortField("length", 5),
+                   ShortField("length", None),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py]: Why are you changing this value?

File: scapy/contrib/gtp_v2.py [hunk 4]
```diff
@@ -633,10 +675,21 @@ class IE_APN(gtp.IE_Base):
...
+                   BitField("instance", 0, 4),
+                   StrLenField("Bearer TFT", "",
+                               length_from=lambda x: x.length)]
+
+
 class IE_AMBR(gtp.IE_Base):
     name = "IE AMBR"
     fields_desc = [ByteEnumField("ietype", 72, IEType),
-                   ShortField("length", 12),
+                   ShortField("length", 8),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py]: This looks incorrect.

File: scapy/contrib/gtp_v2.py [hunk 5]
```diff
@@ -646,8 +699,7 @@ class IE_AMBR(gtp.IE_Base):
 class IE_MSISDN(gtp.IE_Base):
     name = "IE MSISDN"
     fields_desc = [ByteEnumField("ietype", 76, IEType),
-                   FieldLenField("length", None, length_of="digits",
-                                 adjust=lambda pkt, x: x + 4, fmt="H"),
+                   ShortField("length", None),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py]: Why are you changing this value?

File: scapy/contrib/gtp_v2.py [hunk 6]
```diff
@@ -1068,14 +1184,24 @@ class IE_SelectionMode(gtp.IE_Base):
 
 class IE_MMBR(gtp.IE_Base):
     name = "IE Max MBR/APN-AMBR (MMBR)"
-    fields_desc = [ByteEnumField("ietype", 72, IEType),
+    fields_desc = [ByteEnumField("ietype", 161, IEType),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py]: Why are you changing this value?

File: scapy/contrib/gtp_v2.py (hunk 2 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: @guedou Because As per the specs, this is not fixed length IE. So, we cannot hardcode 6 in the length section.

FYI, this change got introduced recently.
Please take a look at here - https://github.com/secdev/scapy/pull/2491/files
I also commented on it saying this is not a correct value.

File: scapy/contrib/gtp_v2.py (hunk 4 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: @guedou  8 is the correct value. This is a fixed length IE. And the length should represent the length of an element; not an entire IE. BTW, I tested my code against a production level GTPv2 packet and it encodes/decodes fine.

File: scapy/contrib/gtp_v2.py (hunk 5 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: Same comment as - https://github.com/secdev/scapy/pull/2488#discussion_r386762044

File: scapy/contrib/gtp_v2.py (hunk 6 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: Same comment as - https://github.com/secdev/scapy/pull/2488#discussion_r386762044

File: scapy/contrib/gtp_v2.py (hunk 3 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: Same comment as - https://github.com/secdev/scapy/pull/2488#discussion_r386762044

File: scapy/contrib/gtp_v2.py (hunk 2 above)
Comment (p-l-) [reply scapy/contrib/gtp_v2.py]: In that case, is there a reason why you do not use a `FieldLenField`?

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: Okay, let me try that.!

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: Hi @p-l- , I've made the changes as requested. Can you please re-review these, Thanks.!

File: scapy/contrib/gtp_v2.py [hunk 7]
```diff
@@ -456,10 +478,28 @@ class IE_ULI(gtp.IE_Base):
 }
 
 
+class IE_UCI(gtp.IE_Base):
+    name = "IE UCI"
+    fields_desc = [ByteEnumField("ietype", 145, IEType),
+                   FieldLenField("length", None, length_of="UCI",
```
Comment (gpotter2) [inline scapy/contrib/gtp_v2.py:484]: Does this work ?! `length_of` expects a field name.5)343 is no field names "UCI". same everywhere

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:484]: Yes, it does.
In fact, I've also added tests to verify this functionality:
```
= IE_UCI, dissection
h = "fe1d70fa717ceeeeeeeeeeee080045000127a4f500003c11e9aec0a8ee80c0a87f50084b23a301131aa1482001070000000001020f009100080021f3540000001602"
gtp = Ether(hex_bytes(h))
ie = gtp.IE_list[0]
ie.CSG_ID == 22 and ie.AccessMode == 0 and ie.LCSG == 1 and ie.CMI == 0 and ie.MCC == b'123' and ie.MNC == b'45'

= IE_UCI, basic instantiation
ie = IE_UCI(ietype='UCI', length=8, CR_flag=0, instance=0, MCC=b'123', MNC=b'45', SPARE=0, CSG_ID=22, AccessMode=0, LCSG=1, CMI=0)
ie.ietype == 145 and ie.CSG_ID == 22 and ie.AccessMode == 0 and ie.LCSG == 1 and ie.CMI == 0 and ie.MCC == b'123' and ie.MNC == b'45'
```

Comment (gpotter2) [reply scapy/contrib/gtp_v2.py:484]: You're not building it.
what about
```
ie = IE_UCI(ietype='UCI', length=8, CR_flag=0, instance=0, MCC=b'123', MNC=b'45', SPARE=0, CSG_ID=22, AccessMode=0, LCSG=1, CMI=0)
bytes(or)
```

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:484]: You mean like this?
```
Python 3.6.8 (default, Aug  7 2019, 17:28:10) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os, binascii
>>> import scapy
>>> from scapy.all import *
>>> from scapy.contrib.gtp_v2 import *
>>> ie = IE_UCI(ietype='UCI', length=8, CR_flag=0, instance=0, MCC=b'123', MNC=b'45', SPARE=0, CSG_ID=22, AccessMode=0, LCSG=1, CMI=0)
>>> ie
<IE_UCI  ietype=UCI length=8 CR_flag=0 instance=0 MCC='123' MNC='45' SPARE=0 CSG_ID=22 AccessMode=0 SPARE=0 LCSG=1 CMI=0 |>
>>> gtpv2_pkt = Ether() / IP(dst='10.15.15.61') / UDP(sport=2123, dport=2123) / GTPHeader(seq=1, P=0, T=0, gtp_type=32, teid=1, length=16) / GTPV2CreateSessionRequest(IE_list=[ie])
>>> gtpv2_pkt
<Ether  type=IPv4 |<IP  frag=0 proto=udp dst=10.15.15.61 |<UDP  sport=gtp_control dport=gtp_control |<GTPHeader  P=0 T=0 gtp_type=create_session_req length=16 seq=1 |<GTPV2CreateSessionRequest  IE_list=[<IE_UCI  ietype=UCI length=8 CR_flag=0 instance=0 MCC='123' MNC='45' SPARE=0 CSG_ID=22 AccessMode=0 SPARE=0 LCSG=1 CMI=0 |>] |>>>>>
>>> 
```
Its working...

Comment (gpotter2) [reply scapy/contrib/gtp_v2.py:484]: Building means converting a packet to bytes. To build it you need to call `bytes()` on it

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:484]: Oh I get it now.
Any suggestions, how I can fix these?
```
>>> gtpv2_pkt.bytes()
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 366, in __getattr__
    fld, v = self.getfield_and_val(attr)
TypeError: 'NoneType' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 366, in __getattr__
    fld, v = self.getfield_and_val(attr)
TypeError: 'NoneType' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 366, in __getattr__
    fld, v = self.getfield_and_val(attr)
TypeError: 'NoneType' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 366, in __getattr__
    fld, v = self.getfield_and_val(attr)
TypeError: 'NoneType' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 366, in __getattr__
    fld, v = self.getfield_and_val(attr)
TypeError: 'NoneType' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 368, in __getattr__
    return self.payload.__getattr__(attr)
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 368, in __getattr__
    return self.payload.__getattr__(attr)
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 368, in __getattr__
    return self.payload.__getattr__(attr)
  [Previous line repeated 2 more times]
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 366, in __getattr__
    fld, v = self.getfield_and_val(attr)
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 1585, in getfield_and_val
    raise AttributeError(attr)
AttributeError: bytes
```

Comment (gpotter2) [reply scapy/contrib/gtp_v2.py:484]: I meant `bytes(gtpv2_pkt)`

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:484]:
```
>>> bytes(gtpv2_pkt)
b'\xff\xff\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x08\x00E\x00\x000\x00\x01\x00\x00@\x11H%\n\x0f\x0f=\n\x0f\x0f=\x08K\x08K\x00\x1cVb@ \x00\x10\x00\x00\x01\x00\x91\x00\x08\x00!\xf3T\x00\x00\x00\x16\x02'
```

Comment (gpotter2) [reply scapy/contrib/gtp_v2.py:484]: Could you only try `bytes(IE_UCI())`. I expect the computation of the length to be broken

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:484]: you are right; looks like it is broken:
```
>>> bytes(IE_UCI())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 492, in __bytes__
    return self.build()
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 612, in build
    p = self.do_build()
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 594, in do_build
    pkt = self.self_build()
  File "/usr/local/lib/python3.6/site-packages/scapy/packet.py", line 575, in self_build
    p = f.addfield(self, p, val)
  File "/usr/local/lib/python3.6/site-packages/scapy/fields.py", line 150, in addfield
    return s + self.struct.pack(self.i2m(pkt, val))
  File "/usr/local/lib/python3.6/site-packages/scapy/fields.py", line 1465, in i2m
    fld, fval = pkt.getfield_and_val(self.length_of)
TypeError: 'NoneType' object is not iterable
```

File: scapy/contrib/gtp_v2.py [hunk 8]
```diff
@@ -333,8 +355,8 @@ class ULI_SAI(ULI_Field):
     fields_desc = [
         gtp.TBCDByteField("MCC", "", 2),
         gtp.TBCDByteField("MNC", "", 1),
-        ShortField("LAC", 0),
-        ShortField("SAC", 0),
+        BitField("LAC", 0, 16),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py:358]: Why is this change needed?

File: scapy/contrib/gtp_v2.py [hunk 9]
```diff
@@ -345,8 +367,8 @@ class ULI_RAI(ULI_Field):
         # MNC: if the third digit of MCC is 0xf, then the length of
         # MNC is 1 byte
         gtp.TBCDByteField("MNC", "", 1),
-        ShortField("LAC", 0),
-        ShortField("RAC", 0),
+        BitField("LAC", 0, 16),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py:370]: Why is this change needed?

File: scapy/contrib/gtp_v2.py [hunk 10]
```diff
@@ -355,7 +377,7 @@ class ULI_TAI(ULI_Field):
     fields_desc = [
         gtp.TBCDByteField("MCC", "", 2),
         gtp.TBCDByteField("MNC", "", 1),
-        ShortField("TAC", 0),
+        BitField("TAC", 0, 16),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py:380]: Why is this change needed?

File: scapy/contrib/gtp_v2.py [hunk 11]
```diff
@@ -374,15 +396,17 @@ class ULI_LAI(ULI_Field):
     fields_desc = [
         gtp.TBCDByteField("MCC", "", 2),
         gtp.TBCDByteField("MNC", "", 1),
-        ShortField("LAC", 0),
+        BitField("LAC", 0, 16),
```
Comment (guedou) [inline scapy/contrib/gtp_v2.py:399]: Why is this change needed? (and all of the changes related to modifying a valid Scapy field into a BitField)

**Comment by guedou** [comment]
And I made some comments. Could you make sure that all the added packets are tested? For example, `IE_FQCSID` is not. Thanks.

File: scapy/contrib/gtp_v2.py (hunk 8 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:358]: Because, I think this is wrongly encoded. As per 3gpp specs, the encoding is as below:
```
  |   | Bits |  
-- | -- | -- | --
  | Octets | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 |  
  | b | MCC digit 2 | MCC digit 1 |  
  | b+1 | MNC digit 3 | MCC digit 3 |  
  | b+2 | MNC digit 2 | MNC digit 1 |  
  | b+3 to b+4 | Location Area Code (LAC) |  
  | b+5 to b+6 | Service Area Code (SAC) |  
  |   |   |   |   |   |   |   |   |   |  

The Location Area Code (LAC) consists of 2 octets. Bit 8 of Octet b+3 is the most significant bit and bit 1 of Octet b+4 the least significant bit. The coding of the location area code is the responsibility of each administration. Coding using full hexadecimal representation (binary, not ASCII encoding) shall be used.
The Service Area Code (SAC) consists of 2 octets. Bit 8 of Octet b+5 is the most significant bit and bit 1 of Octet b+6 the least significant bit. The SAC is defined by the operator. See 3GPP TS 23.003 [2] clause 12.5 for more information. 
```
So, LAC is 2 bytes and SAC is next 2 bytes.
Hence, I made a change to reflect its correct length to avoid wrong encoding & decoding.

Same comment applies for all these `ULI` fields.

**Comment by UmakantKulkarni** [comment]
> And I made some comments. Could you make sure that all the added packets are tested? For example, `IE_FQCSID` is not. Thanks.

Sure will add the missing tests for the new IEs, Thanks!

File: scapy/contrib/gtp_v2.py (hunk 9 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:370]: Same comment as https://github.com/secdev/scapy/pull/2488/files#r425166992 
Took the reference from existing field - https://github.com/secdev/scapy/blob/master/scapy/contrib/gtp_v2.py#L362

File: scapy/contrib/gtp_v2.py (hunk 10 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:380]: Same comment as https://github.com/secdev/scapy/pull/2488/files#r425166992 
Took the reference from existing field - https://github.com/secdev/scapy/blob/master/scapy/contrib/gtp_v2.py#L362

File: scapy/contrib/gtp_v2.py (hunk 11 above)
Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:399]: Same comment as https://github.com/secdev/scapy/pull/2488/files#r425166992 
Took the reference from existing field - https://github.com/secdev/scapy/blob/master/scapy/contrib/gtp_v2.py#L362

File: scapy/contrib/gtp_v2.py (hunk 8 above)
Comment (gpotter2) [reply scapy/contrib/gtp_v2.py:358]: ShortField is 16 bits. Why is this change useful?
ShortField x 2 = bitField(16) x 2

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:358]: Oh, if ShortField is 2 bytes, then that works; I will back out my changes.
Sorry, I thought its 1 byte.

File: scapy/contrib/gtp_v2.py [hunk 12]
```diff
@@ -657,26 +734,41 @@ class IE_MSISDN(gtp.IE_Base):
 class IE_Indication(gtp.IE_Base):
     name = "IE Indication"
     fields_desc = [ByteEnumField("ietype", 77, IEType),
-                   ShortField("length", None),
+                   FieldLenField("length", 8),
```
Comment (gpotter2) [inline scapy/contrib/gtp_v2.py:737]: Looks good except this field; not sure this works?

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:737]: Yes, worked well so far. Since it has all conditional IEs, if user does not want to set any value, it will be encoded as 0 otherwise 1.
Hence total length of this IE will always be 8.

Comment (gpotter2) [reply scapy/contrib/gtp_v2.py:737]: In that case don't use `FieldLenField` ? Since it's not hooked to any fields..

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py:737]: okay, I will make it as ShortField

Comment (UmakantKulkarni) [reply scapy/contrib/gtp_v2.py]: So, 3gpp specs are little confusing for me:
{code}
For each message the applicable flags of the Indication IE shall be clearly specified in the individual message sub clause. The remaining flags of the Indication IE not so indicated shall be discarded by the receiver.
The receiver shall consider the value of the applicable flags as "0", if the Indication IE is applicable for the message but not included in the message by the sender. 
{code}
Thus, to be on the safer side, I modified it to compute the length automatically based on total length of IE. Verified same:
{code}
>>> ie = IE_Indication(ietype='Indication', PS=1, CPRAI=1)
>>> ie
<IE_Indication  ietype=Indication PS=1 CPRAI=1 |>
>>> ie.PPON 
0
>>> sendp(Ether()/IP(src='10.15.15.61', dst='10.9.221.58')/UDP(sport=2123, dport=2123)/GTPHeader(seq=1, P=0, T=1, gtp_type=36, teid=1, length=20)/GTPV2CreateSessionResponse(IE_list=[ie]), iface='eth0')
.
Sent 1 packets.
>>> a = 'd89ef3da40e2fa163e956dce0800450000340001000040117a290a0f0f3d0a09dd3a084b084b00204fc74824001400000001000001004d0008000008008000000000'
>>> gtp = Ether(hex_bytes(a))
>>> gtp.IE_list[0]
<IE_Indication  ietype=Indication length=8 CR_flag=0 instance=0 DAF=0 DTF=0 HI=0 DFI=0 OI=0 ISRSI=0 ISRAI=0 SGWCI=0 SQCI=0 UIMSI=0 CFSI=0 CRSI=0 PS=1 PT=0 SI=0 MSV=0 RetLoc=0 PBIC=0 SRNI=0 S6AF=0 S4AF=0 MBMDT=0 ISRAU=0 CCRSI=0 CPRAI=1 ARRL=0 PPOFF=0 PPON=0 PPSI=0 CSFBI=0 CLII=0 CPSR=0 NSI=0 UASI=0 DTCI=0 BDWI=0 PSCI=0 PCRI=0 AOSI=0 AOPI=0 ROAAI=0 EPCOSI=0 CPOPCI=0 PMTSMI=0 S11TF=0 PNSI=0 UNACCSI=0 WPMSI=0 5GSNN26=0 REPREFI=0 5GSIWKI=0 EEVRSI=0 LTEMUI=0 LTEMPI=0 ENBCRSI=0 TSPCMI=0 Spare=0 Spare=0 Spare=0 N5GNMI=0 5GCNRS=0 5GCNRI=0 5SRHOI=0 ETHPDN=0 |>
>>> gtp.IE_list[0].length
8
{code}

**Comment by guedou** [comment]
You have flake8 errors: https://github.com/secdev/scapy/pull/2488/checks?check_run_id=690106110 Could you fix them?

**Comment by UmakantKulkarni** [comment]
> You have flake8 errors: https://github.com/secdev/scapy/pull/2488/checks?check_run_id=690106110 Could you fix them?

Done, fixed it.

**Comment by UmakantKulkarni** [comment]
@p-l- 
Thanks for approving.
But, it still says "Merging is blocked". Any idea why?

**Comment by p-l-** [comment]
That was related to failed tests (not connected with your PR).
